app/code/Magento/InventoryElasticsearch/Plugin/CatalogSearch/Model/Indexer/Fulltext/Action/DataProvider/StockedProductFilterByInventoryStock.php

```diff
@@ -0,0 +1,204 @@
+    private $stockConfiguration;
+
+    /**
+     * @var StockIndexTableNameResolverInterface
+     */
+    private $stockIndexTableNameResolver;
+
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @var StockByWebsiteIdResolverInterface
+     */
+    private $stockByWebsiteIdResolver;
+
+    /**
+     * @var StockStatusCriteriaInterfaceFactory
+     */
+    private $stockStatusCriteriaFactory;
+
+    /**
+     * @var StockStatusRepositoryInterface
+     */
+    private $stockStatusRepository;
+
+    /**
+     * @var DefaultStockProviderInterface
+     */
+    private $defaultStockProvider;
+
+    /**
+     * @param Config $config
+     * @param StockConfigurationInterface $stockConfiguration
+     * @param StockIndexTableNameResolverInterface $stockIndexTableNameResolver
+     * @param ResourceConnection $resourceConnection
+     * @param StockByWebsiteIdResolverInterface $stockByWebsiteIdResolver
+     * @param StockStatusCriteriaInterfaceFactory $stockStatusCriteriaFactory
+     * @param StockStatusRepositoryInterface $stockStatusRepository
+     * @param DefaultStockProviderInterface $defaultStockProvider
+     */
+    public function __construct(
+        Config $config,
+        StockConfigurationInterface $stockConfiguration,
+        StockIndexTableNameResolverInterface $stockIndexTableNameResolver,
+        ResourceConnection $resourceConnection,
+        StockByWebsiteIdResolverInterface $stockByWebsiteIdResolver,
+        StockStatusCriteriaInterfaceFactory $stockStatusCriteriaFactory,
+        StockStatusRepositoryInterface $stockStatusRepository,
+        DefaultStockProviderInterface $defaultStockProvider
+    ) {
+        $this->config = $config;
+        $this->stockConfiguration = $stockConfiguration;
+        $this->stockIndexTableNameResolver = $stockIndexTableNameResolver;
+        $this->resourceConnection = $resourceConnection;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+        $this->stockStatusCriteriaFactory = $stockStatusCriteriaFactory;
+        $this->stockStatusRepository = $stockStatusRepository;
+        $this->defaultStockProvider = $defaultStockProvider;
+    }
+
+    /**
+     * Filter out of stock options for configurable product.
+     *
+     * @param DataProvider $dataProvider
+     * @param array $indexData
+     * @param array $productData
+     * @param int $storeId
+     * @return array
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforePrepareProductIndex(
+        DataProvider $dataProvider,
+        array $indexData,
+        array $productData,
+        int $storeId
+    ): array {
+        if ($this->config->isElasticsearchEnabled() && !$this->stockConfiguration->isShowOutOfStock($storeId)) {
+            $productIds = array_keys($indexData);
+            $stockStatuses = [];
+
+            foreach ($this->getProductIdsByWebsiteIds($productIds) as $websiteId => $productIdsStr) {
```

Comment:
we have to resolve Website by $storeId and then by Website (which in MSI plays a role of Sales Channel) we could resolve corresponding Stock

Thus, for $stockId you would have ONE and ONLY Stock